cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -491,20 +492,39 @@ private void addWorkflowNodeState(ProgramRunId programRunId, Map<String, String>
     ApplicationId appId = programRunId.getParent().getParent();
     ProgramRunId workflowRunId = appId.workflow(workflowName).run(workflowRun);
 
-    WorkflowNodeStateDetail nodeStateDetail = new WorkflowNodeStateDetail(workflowNodeId,
-                                                                          ProgramRunStatus.toNodeStatus(status),
-                                                                          programRunId.getRun(), failureCause);
-    List<Field<?>> fields = getWorkflowPrimaryKeys(workflowRunId, nodeStateDetail.getNodeId());
-    writeToStructuredTableWithPrimaryKeys(fields, nodeStateDetail, getWorkflowNodeStateTable(),
-                                          StoreDefinition.AppMetadataStore.NODE_STATE_DATA);
-
     // Get the run record of the Workflow which started this program
     List<Field<?>> runRecordFields = getProgramRunInvertedTimeKey(TYPE_RUN_RECORD_ACTIVE, workflowRunId,
                                                                   RunIds.getTime(workflowRun, TimeUnit.SECONDS));
 
-    Optional<StructuredRow> row = getRunRecordsTable().read(runRecordFields);
-    if (row.isPresent()) {
-      RunRecordDetail record = deserializeRunRecordMeta(row.get());
+    RunRecordDetail record = getRunRecordsTable().read(runRecordFields)
+      .map(AppMetadataStore::deserializeRunRecordMeta)
+      .orElse(null);
+
+    // If the workflow is gone, just ignore the update
+    if (record == null) {
+      return;
+    }
+
+    List<Field<?>> primaryKeys = getWorkflowPrimaryKeys(workflowRunId, workflowNodeId);
+    WorkflowNodeStateDetail nodeState = getWorkflowNodeStateTable().read(primaryKeys)
+      .map(r -> r.getString(StoreDefinition.AppMetadataStore.NODE_STATE_DATA))
+      .map(f -> GSON.fromJson(f, WorkflowNodeStateDetail.class))
+      .orElse(null);
+
+    // Update the workflow state table if
+    // - the program state is STARTING (if there is an existing record,
+    //                                  this means this is from a latter attempt of the same node)
+    // - the node state does not exist or
+    // - the program runId is the same as the existing workflow state
+    if (status == ProgramRunStatus.STARTING
```

Comment:
hmm.. not quite sure how this will fix the retry issue.

so when the latter attempt of the same node starts, it still override the former attempt node state ? (because the initially the latter attempt will be starting, and node run id is not part of the primary key, instead it's within node state data)

Reply:
The workflow node state table doesn't affect the program status. It is only for powering the workflow node state API https://cdap.atlassian.net/wiki/spaces/DOCS/pages/477790711/Workflow+Microservices#Workflow-State

That API only cares about the latest status of a node inside a workflow.